### PR TITLE
Split SymNode into its own file

### PR DESCRIPTION
**Pull request by lezcano**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #112038
* __->__ #112037
* #112036
* #112035

This PR:

- Moves TrueDiv, LShift, RShift, IsNonOverlappingAndDenseIndicator to `_sympy.functions.py`
- Moves SymNode to `fx.experimental.sym_node`.
  - This file does not have any SymPy dependencies at import time
  - It installs the magic methods in Sym{Bool,Int,Float}.
  - N.b. With this split, we may be able to move Sym{Bool,Int,Float} to this file, and remove quite a few of the hacks around these classes
- Imports `sym_node` in `torch/__init__.py` rather than the whole `symbolic_shapes.py`.
  This breaks the import-time dependency between torch and SymPy

cc @albanD @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler

### CODE REVIEW

File: torch/fx/experimental/sym_node.py [hunk 1]
```diff
@@ -0,0 +1,1094 @@
...
+    sym_ite,
+    sym_max,
+    sym_min,
+    sym_not,
+    SymBool,
+    SymFloat,
+    SymInt,
+)
+
+from torch.fx.experimental.sym_dispatch_mode import sym_function_mode
```
Comment (mikaylagawarecki) [inline torch/fx/experimental/sym_node.py:34]: There is a small oops here :)
```suggestion
from torch.fx.experimental._sym_dispatch_mode import sym_function_mode
```

Comment (lezcano) [reply torch/fx/experimental/sym_node.py]: oops. Rebasing is difficult!

File: test/allowlist_for_publicAPI.json [hunk 2]
```diff
@@ -1967,6 +1967,51 @@
     "Transformer",
     "Tuple"
   ],
+  "torch.fx.experimental.sym_dispatch_mode": [
+    "sym_function_mode",
+    "set_sym_function_mode"
+  ],
+  "torch.fx.experimental.sym_node": [
+    "SymNode",
```
Comment (peterbell10) [inline test/allowlist_for_publicAPI.json:1975]: I take it these are all just shuffling around, not actually new?

File: torch/fx/experimental/sym_node.py [hunk 3]
```diff
@@ -0,0 +1,1094 @@
...
+    if method in magic_methods_on_submodule:
+        op = getattr(torch.fx.experimental.sym_node, method_attr)
+    elif method in magic_methods_on_math:
+        op = getattr(math, method_attr)
+    else:
+        op = getattr(operator, method_attr)
+    return op
+
+
+# TODO Continue from here. SymNode is going to be tricky...
```
Comment (peterbell10) [inline torch/fx/experimental/sym_node.py:745]: Leftover TODO?

File: test/allowlist_for_publicAPI.json (hunk 2 above)
Comment (lezcano) [reply test/allowlist_for_publicAPI.json:1975]: Yep, just good ol' shuffling + adding the new aux functions I created like `sympy_eq`.

File: torch/fx/experimental/sym_node.py [hunk 4]
```diff
@@ -0,0 +1,1094 @@
...
+    else:
+        method_attr = method
+
+    def binary_magic_impl(self, other):
+        from torch.fx.experimental.symbolic_shapes import (
+            _handle_sym_dispatch,
+            safe_expand,
+            to_node,
+            wrap_node,
+        )
```
Comment (peterbell10) [inline torch/fx/experimental/sym_node.py:760]: Any reason all of these couldn't be moved to this file?

File: test/allowlist_for_publicAPI.json (hunk 2 above)
Comment (peterbell10) [reply test/allowlist_for_publicAPI.json:1975]: Call the function `_sympy_eq` so they aren't public?

File: torch/fx/experimental/sym_node.py (hunk 3 above)
Comment (lezcano) [reply torch/fx/experimental/sym_node.py:745]: Yes, this was my lunch break lol

File: torch/fx/experimental/sym_node.py (hunk 4 above)
Comment (lezcano) [reply torch/fx/experimental/sym_node.py:760]: Moved to_node and wrap_node. I'm leaving safe_expand in the other file as it feels like it belongs in there.

**Comment by lezcano** [comment]
Reviews addressed

**Review by peterbell10 — APPROVED** [review]

**Pull request closed without merging**